### DISCUSSION TITLE
Can I request a formula to be built?

### DISCUSSION THREAD

**Discussion by singularitti** [Homebrew/discussions#4334] [Writing Formulae/Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.6-147-gb3684e5
ORIGIN: https://github.com/Homebrew/brew
HEAD: b3684e55a7a0090e50d975c347961cb57a5ffcb5
Last commit: 12 hours ago
Core tap: N/A
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: code
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

It is my first time requesting a formula here, so I don't think this is the right place. I found an interesting software called [`codon`](https://github.com/exaloop/codon), and it can be built on macOS. I would appreciate it very much if someone could help me build it.

**Comment by SMillerDev** [comment]
https://docs.brew.sh/Adding-Software-to-Homebrew

**Comment by carlocab** [comment]
Their [build instructions](https://docs.exaloop.io/codon/advanced/build) are pretty straightforward. It looks like you can just lift the `install` method from `halide` or `enzyme`.

However, I'm not sure this is suitable for inclusion into Homebrew/core. They vendor multiple dependencies that are provided by Homebrew. See https://github.com/exaloop/codon/blob/2d20e1b35879bd97acec1c0ec3835ef3919dd32f/cmake/deps.cmake.

I suggest either creating a formula in your own tap and/or asking upstream (or sending a PR) to support using system-provided versions of the libraries they vendor. The particularly important ones are `zlib`, `xz`, `bz2`, `bdwgc`, and `re2`. But support for the rest would be better too.

**Reply by gromgit** [reply]
Or see [here](https://github.com/gromgit/homebrew-dev/blob/main/Formula/codon.rb) for a worked example of How Formulae Are Made, specifically how to map Codon source build instructions to Homebrew formula DSL statements.

Or if you just want to use the darn thing, `brew install gromgit/dev/codon`. However:
1. I update everything in `gromgit/dev` as and when I feel like it, so no promises.
2. I'll only contribute this formula to `homebrew/core` when upstream fixes all of @carlocab's concerns. (It looks a lot like upstream thought "I'm providing binaries, plus users will be happy to build everything from scratch, so I don't have to tell them to install anything"...conveniently forgetting `cmake` and `ninja`.)
3. I've included an ARM Monterey bottle, and _may_ include one or two more Intel bottles...when they finish building. If your machine isn't ARM Monterey, you've just triggered a long build session, because upstream said "build everything from scratch...including LLVM 15".

**Reply by carlocab** [reply]
I didn't even notice they used an LLVM fork, because of course they are. That's an even bigger dealbreaker to me -- we have enough formulae building their own copies of LLVM in Homebrew/core.
